src/main/java/com/es/phoneshop/service/DefaultCartService.java

```diff
@@ -0,0 +1,76 @@
+package com.es.phoneshop.service;
+
+import com.es.phoneshop.model.cart.Cart;
+import com.es.phoneshop.model.cart.CartItem;
+import com.es.phoneshop.model.cart.exception.OutOfStockException;
+import com.es.phoneshop.model.product.Product;
+
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class DefaultCartService implements CartService{
+    private static final String CART_SESSION_ATTRIBUTE = DefaultCartService.class.getName() + ".cart";
+    private static final Object instanceLock = new Object();
+    private static volatile DefaultCartService instance;
+
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
```

Comment:
i don't like lockers here, the lecturer suggested to use  session lock here, but i found only this 
https://stackoverflow.com/questions/9802165/is-synchronization-within-an-httpsession-feasible
but what stopped me, is calculating session lock. Is it good solution or not?

Reply:
It's really enough to use synchronized method/block here.
Locks in this lab are mostly to get acquainted with various synchronization techniques.